test/controllers/VerifiedEmailControllerSpec.scala

```diff
@@ -229,14 +229,125 @@ class VerifiedEmailControllerSpec extends SpecBase with MockAuthConnector {
     }
   }
 
+  "retrieveVerifiedEmailThirdParty" should {
+
+    "return Not Found if no data is found in the cache and SUB09 returns no email" in new Setup {
+      when(mockEmailRepository.get(any())).thenReturn(Future.successful(None))
+      when(mockSubscriptionInfoService.getSubscriberInformation(any())).thenReturn(Future.successful(None))
+
+      val request: FakeRequest[AnyContentAsJson] = FakeRequest(POST, retrieveVerifiedEmailThirdPartyRoute).withJsonBody(
+        Json.obj("eori" -> TEST_EORI_VALUE)
+      )
+
+      running(app) {
+        val result = route(app, request).value
+        status(result) mustBe NOT_FOUND
+      }
+    }
+
+    "return the email and not call SUB09 if the data is stored in the cache" in new Setup {
+      when(mockEmailRepository.get(any()))
+        .thenReturn(Future.successful(Some(NotificationEmail(testAddress, testTime, None))))
+
+      when(mockEmailRepository.set(any(), any())).thenReturn(Future.successful(SuccessfulEmail))
+      when(mockSubscriptionInfoService.getSubscriberInformation(any())).thenReturn(Future.successful(None))
+
+      val request: FakeRequest[AnyContentAsJson] = FakeRequest(POST, retrieveVerifiedEmailThirdPartyRoute).withJsonBody(
+        Json.obj("eori" -> TEST_EORI_VALUE)
+      )
+
+      running(app) {
+        val result = route(app, request).value
+        status(result) mustBe OK
+        contentAsJson(result) mustBe Json.obj("address" -> testAddress, "timestamp" -> s"${testTime.toString}Z")
+        verifyNoInteractions(mockSubscriptionInfoService)
+      }
+    }
+
+    "return the email and call SUB09 if the data is not stored in the cache and " +
+      "also store the response into the cache" in new Setup {
+        when(mockEmailRepository.get(any()))
+          .thenReturn(Future.successful(None))
+          .thenReturn(Future.successful(Some(NotificationEmail(testAddress, testTime, None))))
```

Comment:
Is it (second thenReturn) on purpose? I think we can remove this. Tests are working fine after removal